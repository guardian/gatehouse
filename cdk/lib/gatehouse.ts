import {GuPlayApp} from '@guardian/cdk';
import {AccessScope} from '@guardian/cdk/lib/constants/access';
import type {GuStackProps} from '@guardian/cdk/lib/constructs/core';
import {GuDistributionBucketParameter, GuStack, GuStringParameter} from '@guardian/cdk/lib/constructs/core';
import {GuCname} from '@guardian/cdk/lib/constructs/dns';
import {GuPolicy, ReadParametersByName} from '@guardian/cdk/lib/constructs/iam';
import type {App} from 'aws-cdk-lib';
import {Duration} from 'aws-cdk-lib';
import {InstanceClass, InstanceSize, InstanceType, SecurityGroup} from 'aws-cdk-lib/aws-ec2';
import {Effect, PolicyStatement} from 'aws-cdk-lib/aws-iam';
import {ParameterDataType, ParameterTier, StringParameter} from 'aws-cdk-lib/aws-ssm';
import * as xray from 'aws-cdk-lib/aws-xray';

export interface GatehouseStackProps extends GuStackProps {
    domainName: string;
}

export class Gatehouse extends GuStack {
    constructor(
        scope: App,
        id: string,
        props: GatehouseStackProps,
    ) {
        super(scope, id, props);

        const ec2App = 'gatehouse';

        const distBucket = GuDistributionBucketParameter.getInstance(this).valueAsString;

        const artifactPath = [distBucket, this.stack, this.stage, ec2App, `${ec2App}.deb`].join('/');

        const readAppSsmParamsPolicy = new GuPolicy(this, 'ReadAppSsmParamsPolicy', {
            statements: [new ReadParametersByName(this, {app: ec2App})]
        })

        // See https://aws-otel.github.io/docs/setup/permissions
        const xrayTelemetryPolicy = new GuPolicy(this, 'XrayTelemetryPolicy', {
            statements: [new PolicyStatement({
                effect: Effect.ALLOW,
                actions: [
                    'logs:PutLogEvents',
                    'logs:CreateLogGroup',
                    'logs:CreateLogStream',
                    'logs:DescribeLogStreams',
                    'logs:DescribeLogGroups',
                    'logs:PutRetentionPolicy',
                    'xray:PutTraceSegments',
                    'xray:PutTelemetryRecords',
                    'xray:GetSamplingRules',
                    'xray:GetSamplingTargets',
                    'xray:GetSamplingStatisticSummaries',
                ],
                resources: ['*']
            })]
        })

        const rdsSecurityGroupId = new GuStringParameter(this, 'rdsSecurityGroupId', {
            fromSSM: true,
            default: `/${this.stage}/${this.stack}/${ec2App}/rdsSecurityGroup/id`,
            description: 'ID of database security group.',
        });

        const app = new GuPlayApp(this, {
            app: ec2App,
            instanceType: InstanceType.of(InstanceClass.T4G, InstanceSize.MICRO),
            access: {scope: AccessScope.PUBLIC},
            userData: [
                '#!/bin/bash -ev',

                // See https://github.com/aws-observability/aws-otel-collector/blob/main/docs/developers/linux-rpm-demo.md
                '# Install X-Ray Collector',
                'wget -P /tmp https://aws-otel-collector.s3.amazonaws.com/ubuntu/arm64/latest/aws-otel-collector.deb',
                'dpkg -i /tmp/aws-otel-collector.deb',
                'cat << EOF > /opt/aws/aws-otel-collector/etc/config.yaml',
                'receivers:',
                '  otlp:',
                '    protocols:',
                '      grpc:',
                'processors:',
                '  resourcedetection/ec2:',
                '    detectors:',
                '      - ec2',
                '    ec2:',
                '      tags:',
                '        - Stage',
                '    timeout: 2s',
                '    override: false',
                '  memory_limiter:',
                '    check_interval: 1s',
                '    limit_mib: 20',
                '  batch/traces:',
                '    timeout: 1s',
                '    send_batch_size: 50',
                'exporters:',
                '  awsxray:',
                // '    index_all_attributes: true',
                '    indexed_attributes:',
                '      - otel_resource_ec2_tag_Stage',
                'extensions:',
                '  awsproxy:',
                'service:',
                '  extensions:',
                '    - awsproxy',
                '  pipelines:',
                '    traces:',
                '      receivers:',
                '        - otlp',
                '      processors:',
                '        - resourcedetection/ec2',
                '        - memory_limiter',
                '        - batch/traces',
                '      exporters:',
                '        - awsxray',
                'EOF',
                'echo "loggingLevel=DEBUG" | sudo tee -a /opt/aws/aws-otel-collector/etc/extracfg.txt',
                'sudo /opt/aws/aws-otel-collector/bin/aws-otel-collector-ctl -a start',

                // See https://aws-otel.github.io/docs/getting-started/java-sdk/auto-instr
                '# Install X-Ray Agent',
                'sudo mkdir /opt/aws-opentelemetry-agent',
                'chmod +rx /opt/aws-opentelemetry-agent',
                'wget -P /opt/aws-opentelemetry-agent https://github.com/aws-observability/aws-otel-java-instrumentation/releases/latest/download/aws-opentelemetry-agent.jar',

                '# Install app',
                `aws --region ${props.env?.region} s3 cp s3://${artifactPath} /tmp/${ec2App}.deb`,
                `dpkg -i /tmp/${ec2App}.deb`,
            ].join('\n'),
            certificateProps: {
                domainName: props.domainName,
            },
            monitoringConfiguration: {noMonitoring: true},
            scaling: {
                minimumInstances: 1,
                maximumInstances: 2,
            },
            applicationLogging: {
                enabled: true,
                systemdUnitName: 'gatehouse'
            },
            imageRecipe: 'arm-identity-base-jammy-java21-cdk-base',
            roleConfiguration: {
                additionalPolicies: [
                    readAppSsmParamsPolicy,
                    xrayTelemetryPolicy,
                ],
            },
        });

        app.autoScalingGroup.connections.addSecurityGroup(
            SecurityGroup.fromSecurityGroupId(this, 'rdsSecurityGroup', rdsSecurityGroupId.valueAsString, {
                mutable: false
            }));

        // This parameter is used by https://github.com/guardian/waf
        new StringParameter(this, 'AlbSsmParam', {
            parameterName: `/infosec/waf/services/${this.stage}/gatehouse-alb-arn`,
            description: `The ARN of the ALB for identity-${this.stage}-gatehouse. N.B. This parameter is created via CDK.`,
            simpleName: false,
            stringValue: app.loadBalancer.loadBalancerArn,
            tier: ParameterTier.STANDARD,
            dataType: ParameterDataType.TEXT,
        });

        new GuCname(this, 'EC2AppDNS', {
            app: ec2App,
            ttl: Duration.hours(1),
            domainName: props.domainName,
            resourceRecord: app.loadBalancer.loadBalancerDnsName,
        });

        new xray.CfnGroup(this, 'CodeXrayGroup', {
            groupName: 'CODE',
            filterExpression: 'annotation.otel_resource_ec2_tag_Stage = "CODE"',
            insightsConfiguration: {insightsEnabled: true},
        });
    }
}
