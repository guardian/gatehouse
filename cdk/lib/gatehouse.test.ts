import {App} from 'aws-cdk-lib';
import {Template} from 'aws-cdk-lib/assertions';
import {Gatehouse} from './gatehouse';

describe('The Gatehouse stack', () => {
    it('matches the snapshot', () => {
        const app = new App();
        const stack = new Gatehouse(app, 'gatehouse-TEST', {
            stack: 'identity',
            stage: 'TEST',
            domainName: 'id.test.dev-guardianapis.com',
            env: {region: 'eu-west-1'},
        });
        const template = Template.fromStack(stack);
        expect(template.toJSON()).toMatchSnapshot();
    });
});
