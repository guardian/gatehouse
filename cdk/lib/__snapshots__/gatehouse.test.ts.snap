// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Gatehouse stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuPolicy",
      "GuPolicy",
      "GuStringParameter",
      "GuVpcParameter",
      "GuSubnetListParameter",
      "GuSubnetListParameter",
      "GuPlayApp",
      "GuCertificate",
      "GuInstanceRole",
      "GuDescribeEC2Policy",
      "GuLoggingStreamNameParameter",
      "GuLogShippingPolicy",
      "GuGetDistributablePolicy",
      "GuParameterStoreReadPolicy",
      "GuAmiParameter",
      "GuHttpsEgressSecurityGroup",
      "GuWazuhAccess",
      "GuAutoScalingGroup",
      "GuApplicationLoadBalancer",
      "GuApplicationTargetGroup",
      "GuHttpsApplicationListener",
      "GuCname",
    ],
    "gu:cdk:version": "TEST",
  },
  "Outputs": {
    "LoadBalancerGatehouseDnsName": {
      "Description": "DNS entry for LoadBalancerGatehouse",
      "Value": {
        "Fn::GetAtt": [
          "LoadBalancerGatehouse6031E987",
          "DNSName",
        ],
      },
    },
  },
  "Parameters": {
    "AMIGatehouse": {
      "Description": "Amazon Machine Image ID for the app gatehouse. Use this in conjunction with AMIgo to keep AMIs up to date.",
      "Type": "AWS::EC2::Image::Id",
    },
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "LoggingStreamName": {
      "Default": "/account/services/logging.stream.name",
      "Description": "SSM parameter containing the Name (not ARN) on the kinesis stream",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "VpcId": {
      "Default": "/account/vpc/primary/id",
      "Description": "Virtual Private Cloud to run EC2 instances within. Should NOT be the account default VPC.",
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::VPC::Id>",
    },
    "gatehousePrivateSubnets": {
      "Default": "/account/vpc/primary/subnets/private",
      "Description": "A list of private subnets",
      "Type": "AWS::SSM::Parameter::Value<List<AWS::EC2::Subnet::Id>>",
    },
    "gatehousePublicSubnets": {
      "Default": "/account/vpc/primary/subnets/public",
      "Description": "A list of public subnets",
      "Type": "AWS::SSM::Parameter::Value<List<AWS::EC2::Subnet::Id>>",
    },
    "rdsSecurityGroupId": {
      "Default": "/TEST/identity/gatehouse/rdsSecurityGroup/id",
      "Description": "ID of database security group.",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AlbSsmParam485C1D52": {
      "Properties": {
        "DataType": "text",
        "Description": "The ARN of the ALB for identity-TEST-gatehouse. N.B. This parameter is created via CDK.",
        "Name": "/infosec/waf/services/TEST/gatehouse-alb-arn",
        "Tags": {
          "Stack": "identity",
          "Stage": "TEST",
          "gu:cdk:version": "TEST",
          "gu:repo": "guardian/gatehouse",
        },
        "Tier": "Standard",
        "Type": "String",
        "Value": {
          "Ref": "LoadBalancerGatehouse6031E987",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "AutoScalingGroupGatehouseASG9482EAEF": {
      "Properties": {
        "HealthCheckGracePeriod": 120,
        "HealthCheckType": "ELB",
        "LaunchTemplate": {
          "LaunchTemplateId": {
            "Ref": "identityTESTgatehouseBDA14DB7",
          },
          "Version": {
            "Fn::GetAtt": [
              "identityTESTgatehouseBDA14DB7",
              "LatestVersionNumber",
            ],
          },
        },
        "MaxSize": "2",
        "MinSize": "1",
        "Tags": [
          {
            "Key": "App",
            "PropagateAtLaunch": true,
            "Value": "gatehouse",
          },
          {
            "Key": "gu:cdk:version",
            "PropagateAtLaunch": true,
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "PropagateAtLaunch": true,
            "Value": "guardian/gatehouse",
          },
          {
            "Key": "LogKinesisStreamName",
            "PropagateAtLaunch": true,
            "Value": {
              "Ref": "LoggingStreamName",
            },
          },
          {
            "Key": "Stack",
            "PropagateAtLaunch": true,
            "Value": "identity",
          },
          {
            "Key": "Stage",
            "PropagateAtLaunch": true,
            "Value": "TEST",
          },
          {
            "Key": "SystemdUnit",
            "PropagateAtLaunch": true,
            "Value": "gatehouse.service",
          },
        ],
        "TargetGroupARNs": [
          {
            "Ref": "TargetGroupGatehouseEB93E2A5",
          },
        ],
        "VPCZoneIdentifier": {
          "Ref": "gatehousePrivateSubnets",
        },
      },
      "Type": "AWS::AutoScaling::AutoScalingGroup",
    },
    "CertificateGatehouse0CF1A666": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "DomainName": "id.test.dev-guardianapis.com",
        "Tags": [
          {
            "Key": "App",
            "Value": "gatehouse",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/gatehouse",
          },
          {
            "Key": "Name",
            "Value": "gatehouse-TEST/CertificateGatehouse",
          },
          {
            "Key": "Stack",
            "Value": "identity",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "ValidationMethod": "DNS",
      },
      "Type": "AWS::CertificateManager::Certificate",
      "UpdateReplacePolicy": "Retain",
    },
    "DescribeEC2PolicyFF5F9295": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "autoscaling:DescribeAutoScalingInstances",
                "autoscaling:DescribeAutoScalingGroups",
                "ec2:DescribeTags",
                "ec2:DescribeInstances",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "describe-ec2-policy",
        "Roles": [
          {
            "Ref": "InstanceRoleGatehouse075DE24E",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "EC2AppDNS": {
      "Properties": {
        "Name": "id.test.dev-guardianapis.com",
        "RecordType": "CNAME",
        "ResourceRecords": [
          {
            "Fn::GetAtt": [
              "LoadBalancerGatehouse6031E987",
              "DNSName",
            ],
          },
        ],
        "Stage": "TEST",
        "TTL": 3600,
      },
      "Type": "Guardian::DNS::RecordSet",
    },
    "GetDistributablePolicyGatehouse34FFC5AF": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:s3:::",
                    {
                      "Ref": "DistributionBucketName",
                    },
                    "/identity/TEST/gatehouse/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GetDistributablePolicyGatehouse34FFC5AF",
        "Roles": [
          {
            "Ref": "InstanceRoleGatehouse075DE24E",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GuHttpsEgressSecurityGroupGatehouseD78E25BB": {
      "Properties": {
        "GroupDescription": "Allow all outbound HTTPS traffic",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound HTTPS traffic",
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "gatehouse",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/gatehouse",
          },
          {
            "Key": "Stack",
            "Value": "identity",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": {
          "Ref": "VpcId",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "GuHttpsEgressSecurityGroupGatehousefromgatehouseTESTLoadBalancerGatehouseSecurityGroup226CEBB690008C0B659F": {
      "Properties": {
        "Description": "Load balancer to target",
        "FromPort": 9000,
        "GroupId": {
          "Fn::GetAtt": [
            "GuHttpsEgressSecurityGroupGatehouseD78E25BB",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": {
          "Fn::GetAtt": [
            "LoadBalancerGatehouseSecurityGroupDD25BC00",
            "GroupId",
          ],
        },
        "ToPort": 9000,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "GuLogShippingPolicy981BFE5A": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "kinesis:Describe*",
                "kinesis:Put*",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:kinesis:eu-west-1:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":stream/",
                    {
                      "Ref": "LoggingStreamName",
                    },
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GuLogShippingPolicy981BFE5A",
        "Roles": [
          {
            "Ref": "InstanceRoleGatehouse075DE24E",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "InstanceRoleGatehouse075DE24E": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ec2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AmazonSSMManagedInstanceCore",
              ],
            ],
          },
        ],
        "Path": "/",
        "Tags": [
          {
            "Key": "App",
            "Value": "gatehouse",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/gatehouse",
          },
          {
            "Key": "Stack",
            "Value": "identity",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ListenerGatehouseDDE3C56A": {
      "Properties": {
        "Certificates": [
          {
            "CertificateArn": {
              "Ref": "CertificateGatehouse0CF1A666",
            },
          },
        ],
        "DefaultActions": [
          {
            "TargetGroupArn": {
              "Ref": "TargetGroupGatehouseEB93E2A5",
            },
            "Type": "forward",
          },
        ],
        "LoadBalancerArn": {
          "Ref": "LoadBalancerGatehouse6031E987",
        },
        "Port": 443,
        "Protocol": "HTTPS",
      },
      "Type": "AWS::ElasticLoadBalancingV2::Listener",
    },
    "LoadBalancerGatehouse6031E987": {
      "Properties": {
        "LoadBalancerAttributes": [
          {
            "Key": "deletion_protection.enabled",
            "Value": "true",
          },
        ],
        "Scheme": "internet-facing",
        "SecurityGroups": [
          {
            "Fn::GetAtt": [
              "LoadBalancerGatehouseSecurityGroupDD25BC00",
              "GroupId",
            ],
          },
        ],
        "Subnets": {
          "Ref": "gatehousePublicSubnets",
        },
        "Tags": [
          {
            "Key": "App",
            "Value": "gatehouse",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/gatehouse",
          },
          {
            "Key": "Stack",
            "Value": "identity",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Type": "application",
      },
      "Type": "AWS::ElasticLoadBalancingV2::LoadBalancer",
    },
    "LoadBalancerGatehouseSecurityGroupDD25BC00": {
      "Properties": {
        "GroupDescription": "Automatically created Security Group for ELB gatehouseTESTLoadBalancerGatehouseC8325917",
        "SecurityGroupIngress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow from anyone on port 443",
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "gatehouse",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/gatehouse",
          },
          {
            "Key": "Stack",
            "Value": "identity",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": {
          "Ref": "VpcId",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "LoadBalancerGatehouseSecurityGrouptogatehouseTESTGuHttpsEgressSecurityGroupGatehouse8627040E90000595EFE2": {
      "Properties": {
        "Description": "Load balancer to target",
        "DestinationSecurityGroupId": {
          "Fn::GetAtt": [
            "GuHttpsEgressSecurityGroupGatehouseD78E25BB",
            "GroupId",
          ],
        },
        "FromPort": 9000,
        "GroupId": {
          "Fn::GetAtt": [
            "LoadBalancerGatehouseSecurityGroupDD25BC00",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "ToPort": 9000,
      },
      "Type": "AWS::EC2::SecurityGroupEgress",
    },
    "LoadBalancerGatehouseSecurityGrouptogatehouseTESTWazuhSecurityGroup77ADAF5C9000EAF2532A": {
      "Properties": {
        "Description": "Load balancer to target",
        "DestinationSecurityGroupId": {
          "Fn::GetAtt": [
            "WazuhSecurityGroup",
            "GroupId",
          ],
        },
        "FromPort": 9000,
        "GroupId": {
          "Fn::GetAtt": [
            "LoadBalancerGatehouseSecurityGroupDD25BC00",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "ToPort": 9000,
      },
      "Type": "AWS::EC2::SecurityGroupEgress",
    },
    "LoadBalancerGatehouseSecurityGrouptogatehouseTESTrdsSecurityGroup6652430490005CC1E089": {
      "Properties": {
        "Description": "Load balancer to target",
        "DestinationSecurityGroupId": {
          "Ref": "rdsSecurityGroupId",
        },
        "FromPort": 9000,
        "GroupId": {
          "Fn::GetAtt": [
            "LoadBalancerGatehouseSecurityGroupDD25BC00",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "ToPort": 9000,
      },
      "Type": "AWS::EC2::SecurityGroupEgress",
    },
    "ParameterStoreReadGatehouseF4274D8F": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:eu-west-1:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/identity/gatehouse",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:eu-west-1:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/identity/gatehouse/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "parameter-store-read-policy",
        "Roles": [
          {
            "Ref": "InstanceRoleGatehouse075DE24E",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ReadAppSsmParamsPolicy9DC17627": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:eu-west-1:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/identity/gatehouse/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ReadAppSsmParamsPolicy9DC17627",
        "Roles": [
          {
            "Ref": "InstanceRoleGatehouse075DE24E",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TargetGroupGatehouseEB93E2A5": {
      "Properties": {
        "HealthCheckIntervalSeconds": 10,
        "HealthCheckPath": "/healthcheck",
        "HealthCheckProtocol": "HTTP",
        "HealthCheckTimeoutSeconds": 5,
        "HealthyThresholdCount": 5,
        "Port": 9000,
        "Protocol": "HTTP",
        "Tags": [
          {
            "Key": "App",
            "Value": "gatehouse",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/gatehouse",
          },
          {
            "Key": "Stack",
            "Value": "identity",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "TargetGroupAttributes": [
          {
            "Key": "deregistration_delay.timeout_seconds",
            "Value": "30",
          },
          {
            "Key": "stickiness.enabled",
            "Value": "false",
          },
        ],
        "TargetType": "instance",
        "UnhealthyThresholdCount": 2,
        "VpcId": {
          "Ref": "VpcId",
        },
      },
      "Type": "AWS::ElasticLoadBalancingV2::TargetGroup",
    },
    "WazuhSecurityGroup": {
      "Properties": {
        "GroupDescription": "Allow outbound traffic from wazuh agent to manager",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Wazuh event logging",
            "FromPort": 1514,
            "IpProtocol": "tcp",
            "ToPort": 1514,
          },
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Wazuh agent registration",
            "FromPort": 1515,
            "IpProtocol": "tcp",
            "ToPort": 1515,
          },
        ],
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/gatehouse",
          },
          {
            "Key": "Stack",
            "Value": "identity",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": {
          "Ref": "VpcId",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "WazuhSecurityGroupfromgatehouseTESTLoadBalancerGatehouseSecurityGroup226CEBB69000612267A4": {
      "Properties": {
        "Description": "Load balancer to target",
        "FromPort": 9000,
        "GroupId": {
          "Fn::GetAtt": [
            "WazuhSecurityGroup",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": {
          "Fn::GetAtt": [
            "LoadBalancerGatehouseSecurityGroupDD25BC00",
            "GroupId",
          ],
        },
        "ToPort": 9000,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "XrayTelemetryPolicy20B2C572": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutLogEvents",
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:DescribeLogStreams",
                "logs:DescribeLogGroups",
                "logs:PutRetentionPolicy",
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
                "xray:GetSamplingRules",
                "xray:GetSamplingTargets",
                "xray:GetSamplingStatisticSummaries",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "XrayTelemetryPolicy20B2C572",
        "Roles": [
          {
            "Ref": "InstanceRoleGatehouse075DE24E",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "identityTESTgatehouseBDA14DB7": {
      "DependsOn": [
        "InstanceRoleGatehouse075DE24E",
      ],
      "Properties": {
        "LaunchTemplateData": {
          "IamInstanceProfile": {
            "Arn": {
              "Fn::GetAtt": [
                "identityTESTgatehouseProfileBF43149C",
                "Arn",
              ],
            },
          },
          "ImageId": {
            "Ref": "AMIGatehouse",
          },
          "InstanceType": "t4g.micro",
          "MetadataOptions": {
            "HttpTokens": "required",
          },
          "SecurityGroupIds": [
            {
              "Fn::GetAtt": [
                "GuHttpsEgressSecurityGroupGatehouseD78E25BB",
                "GroupId",
              ],
            },
            {
              "Fn::GetAtt": [
                "WazuhSecurityGroup",
                "GroupId",
              ],
            },
            {
              "Ref": "rdsSecurityGroupId",
            },
          ],
          "TagSpecifications": [
            {
              "ResourceType": "instance",
              "Tags": [
                {
                  "Key": "gu:cdk:version",
                  "Value": "TEST",
                },
                {
                  "Key": "gu:repo",
                  "Value": "guardian/gatehouse",
                },
                {
                  "Key": "Name",
                  "Value": "gatehouse-TEST/identity-TEST-gatehouse",
                },
                {
                  "Key": "Stack",
                  "Value": "identity",
                },
                {
                  "Key": "Stage",
                  "Value": "TEST",
                },
              ],
            },
            {
              "ResourceType": "volume",
              "Tags": [
                {
                  "Key": "gu:cdk:version",
                  "Value": "TEST",
                },
                {
                  "Key": "gu:repo",
                  "Value": "guardian/gatehouse",
                },
                {
                  "Key": "Name",
                  "Value": "gatehouse-TEST/identity-TEST-gatehouse",
                },
                {
                  "Key": "Stack",
                  "Value": "identity",
                },
                {
                  "Key": "Stage",
                  "Value": "TEST",
                },
              ],
            },
          ],
          "UserData": {
            "Fn::Base64": {
              "Fn::Join": [
                "",
                [
                  "#!/bin/bash -ev
# Install X-Ray Collector
wget -P /tmp https://aws-otel-collector.s3.amazonaws.com/ubuntu/arm64/latest/aws-otel-collector.deb
dpkg -i /tmp/aws-otel-collector.deb
cat << EOF > /opt/aws/aws-otel-collector/etc/config.yaml
receivers:
  otlp:
    protocols:
      grpc:
processors:
  resourcedetection/ec2:
    detectors:
      - ec2
    ec2:
      tags:
        - Stage
    timeout: 2s
    override: false
  memory_limiter:
    check_interval: 1s
    limit_mib: 20
  batch/traces:
    timeout: 1s
    send_batch_size: 50
exporters:
  awsxray:
    indexed_attributes:
      - otel.resource.ec2.tag.Stage
extensions:
  awsproxy:
service:
  extensions:
    - awsproxy
  pipelines:
    traces:
      receivers:
        - otlp
      processors:
        - resourcedetection/ec2
        - memory_limiter
        - batch/traces
      exporters:
        - awsxray
EOF
sudo /opt/aws/aws-otel-collector/bin/aws-otel-collector-ctl -a start
# Install X-Ray Agent
sudo mkdir /opt/aws-opentelemetry-agent
chmod +rx /opt/aws-opentelemetry-agent
wget -P /opt/aws-opentelemetry-agent https://github.com/aws-observability/aws-otel-java-instrumentation/releases/latest/download/aws-opentelemetry-agent.jar
# Install app
aws --region eu-west-1 s3 cp s3://",
                  {
                    "Ref": "DistributionBucketName",
                  },
                  "/identity/TEST/gatehouse/gatehouse.deb /tmp/gatehouse.deb
dpkg -i /tmp/gatehouse.deb",
                ],
              ],
            },
          },
        },
        "TagSpecifications": [
          {
            "ResourceType": "launch-template",
            "Tags": [
              {
                "Key": "gu:cdk:version",
                "Value": "TEST",
              },
              {
                "Key": "gu:repo",
                "Value": "guardian/gatehouse",
              },
              {
                "Key": "Name",
                "Value": "gatehouse-TEST/identity-TEST-gatehouse",
              },
              {
                "Key": "Stack",
                "Value": "identity",
              },
              {
                "Key": "Stage",
                "Value": "TEST",
              },
            ],
          },
        ],
      },
      "Type": "AWS::EC2::LaunchTemplate",
    },
    "identityTESTgatehouseProfileBF43149C": {
      "Properties": {
        "Roles": [
          {
            "Ref": "InstanceRoleGatehouse075DE24E",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
  },
}
`;
